### PR TITLE
Implement testing for the entire repo

### PR DESCRIPTION
**Pull request by caldwellst**
This is a draft PR just looking at implementing (the hardest) part of the testing, on the `src/email` codebase. This is also tackling a few other open issues at the same time, including:

- Consolidation of modules so that we can import functions fromt he module, not the scripts themselves (basically like using `__init__.py` in Python, just `__init__.R` in this case! Read more [here].(https://klmr.me/box/articles/box.html#nesting-modules).
- Have to work around testing `{box}` modules, which requires some special structure. Read more [here](https://cran.rstudio.com/web/packages/box/vignettes/testing.html).
- We currently streamline some slow pings to the Mailchimp API by loading the data once at the time a module is imported. For instance, see [here](https://github.com/OCHA-DAP/hdx-signals/blob/0a114082346f5fca09cadb8e17c5caf60f0b8394/src/email/mailchimp/custom_segmentation.R#L190) where we load in some lists from the Mailchimp API. No longer can do that because the API is accessed when we load the module (and we have to load the modules to test!). We will move to use `{memoise}`. Not put in this draft PR.

The restructuring will also make things much cleaner in various modules: the indicators themselves will now have a unified structure, much different than how we do now, which will facilitate easier use and easier testing.

### CODE REVIEW

**Comment by zackarno** [comment]
looks cool! I assume not ready for review, but will start looking at the resources you shared as links. 

If I wanted to play around and manually run through some of the tests, what would be the order i'd run things

**Comment by caldwellst** [comment]
If you want to run tests, all you have to do is source the modules that have the lines:

```r
if (is.null(box::name())) {
  box::use(dir/loc/`__tests__`
}
```

**Comment by caldwellst** [comment]
Otherwise, run the `helper-module.R` and `__init__.R` scripts for the relevant directory to ensure relevant modules are loaded and then you can test the tests yourself.

**Comment by zackarno** [comment]
So just to confirm the instructions:

> If you want to run tests, all you have to do is source the modules that have the lines:
> 
> ```r
> if (is.null(box::name())) {
>   box::use(dir/loc/`__tests__`
> }
> ```

 here by "sourcing" the modules you don't actually mean `sourcing` as in `source()` or as in `box::use()` as written in the `helper-module.R` function: `box::use(src/email/components[...])`.  But rather you mean : go to the `_init_.R` file in the parent dir of `_tests_` dir and run that file (i.e `src/email/components/_init_.R`)... i see this does initiate testing which looks like: 

```
#' @export
box::use(
  src/email/components/banner_block[...],
  src/email/components/conditional_merge[...],
  src/email/components/create_template[...],
  src/email/components/email_body[...],
  src/email/components/further_info_block[...],
  src/email/components/image_block[...],
  src/email/components/intro_block[...],
  src/email/components/line_block[...],
  src/email/components/location_block[...],
  src/email/components/missing[...],
  src/email/components/summary_block[...],
  src/email/components/text_block[...]
)

if (is.null(box::name())) {
  box::use(src/email/components/`__tests__`)
}
```


```
✔ | F W  S  OK | Context
✔ |          2 | banner_block                                                                    
✖ | 1        0 | conditional_merge                                                               
─────────────────────────────────────────────────────────────────────────────────────────────────
Error (test-conditional_merge.R:5:1): (code run outside of `test_that()`)
Error in `stub(conditional_merge, "location_codes$iso3_to_regions", "Region A")`: could not find function "stub"
─────────────────────────────────────────────────────────────────────────────────────────────────
✔ |          2 | create_template                                                                 
✖ | 1        0 | email_body                                                                      
─────────────────────────────────────────────────────────────────────────────────────────────────
Error (test-email_body.R:28:3): Runs and returns single string
Error in `local_dir(here$here())`: could not find function "local_dir"
─────────────────────────────────────────────────────────────────────────────────────────────────
✔ |          2 | further_info_block                                                              
✔ |          2 | image_block                                                                     
✔ |          2 | intro_block                                                                     
✔ |          2 | line_block                                                                      
✖ | 1        0 | location_block                                                                  
─────────────────────────────────────────────────────────────────────────────────────────────────
Error (test-location_block.R:9:3): Runs and returns single string
Error in `stub(add_location, "conditional_merge$conditional_merge", function(x) x)`: could not find function "stub"
─────────────────────────────────────────────────────────────────────────────────────────────────
✔ |          5 | missing                                                                         
✔ |          2 | summary_block                                                                   
✔ |          4 | text_block                                                                      

══ Results ══════════════════════════════════════════════════════════════════════════════════════
── Failed tests ─────────────────────────────────────────────────────────────────────────────────
Error (test-conditional_merge.R:5:1): (code run outside of `test_that()`)
Error in `stub(conditional_merge, "location_codes$iso3_to_regions", "Region A")`: could not find function "stub"

Error (test-email_body.R:28:3): Runs and returns single string
Error in `local_dir(here$here())`: could not find function "local_dir"

Error (test-location_block.R:9:3): Runs and returns single string
Error in `stub(add_location, "conditional_merge$conditional_merge", function(x) x)`: could not find function "stub"

[ FAIL 3 | WARN 0 | SKIP 0 | PASS 23 ]
```

By this:

> Otherwise, run the helper-module.R and __init__.R scripts for the relevant directory to ensure relevant modules are loaded and then you can test the tests yourself.


You mean we can manually run  the `helper-module.R` file and then the `_init_.R` file in that same directory as the `helper-module.R` file and our environment will be loaded for testing so that we can walk through tests and or write additional ones. This seems to be the case however the we get the following Error w/ box::export?

```
> box::export()
Error in box::export() : “export” can only be called from inside a module
```

So i guess just skip this line for dev/testing?

**Comment by caldwellst** [comment]
Sorry, didn't respond to your stuff earlier, will do now! Note that the `email` and `utils` folder are now essentially fully tested, although there's a few functions I didn't bother testing in `utils`. 

Your errors are because for `email/components` the `box::use()` calls for `{withr}` and `{mockery} were put in `__init__.R` rather than `the `helper-module.R` module. I've fixed that so they should work now. The result of your explanation is basically correct. `box::export()` doesn't need to be run, it is there because it ensures we designate a script as a module (so it can be called by `box::use()`, and it only needs to be used if you don't have a `#' @export` call (so just in these `tests/__init__.R` scripts).

Otherwise, you can start reviewing some of this testing now, because it's a long and expansive amount of code @martinig94 @zackarno.

**Comment by caldwellst** [comment]
Finalized more of this, including documentation. `email` and `utils` are now fully tested. You will need to check if they are working fully on the remote runs in `.github/workflows/test_signals.yml`. The reason they wouldn't is probably because of missing env variables like `MAILCHIMP_API_KEY` or others tested by `get_env()`. I can't test now as I seemingly am unable to call any actions.

Good luck with extending the testing!

File: src/email/mailchimp/__tests__/test-templates.R [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        result <- mc_add_template(
+          html = "",
+          folder = 1,
+          preview = FALSE
+        )
+
+        expect_equal(result, "1")
+      })
+    })
+})
```
Comment (zackarno) [inline src/email/mailchimp/__tests__/test-templates.R:43]: when running this interactively i get: 

```
── Error: mc_add_template works correctly with local FALSE ─────────────────────
<httptest2_request/rlang_error/error/condition>
Error in `stop_request(req)`: An unexpected request was made:
POST https://us14.api.mailchimp.com/3.0/templates {"name":123,"html":"","folder_id":1}
Expected mock file: us14.api.mailchimp.com/3.0/templates-745a00-POST.*
Backtrace:
     ▆
  1. ├─withr::with_envvar(...)
  2. │ └─base::force(code)
  3. ├─httptest2::with_mock_api(...)
  4. │ └─base::eval.parent(expr)
  5. │   └─base::eval(expr, p)
  6. └─mc_add_template(html = "", folder = 1, preview = FALSE)
  7.   ├─httr2$resp_body_json(...) at email/mailchimp/templates.R:37:5
  8.   │ └─httr2:::check_response(resp)
  9.   │   └─httr2:::is_response(resp)
 10.   └─httr2$req_perform(...)
 11.     └─httptest2 (local) mock(req)
 12.       └─httptest2:::stop_request(req)
 13.         └─rlang::abort(out, mockfile = req$mockfile, class = "httptest2_request")

Error:
! Test failed
Backtrace:
    ▆
 1. ├─testthat::test_that(...)
 2. │ └─withr (local) `<fn>`()
 3. └─reporter$stop_if_needed()
 4.   └─rlang::abort("Test failed", call = NULL)
```

File: src/email/mailchimp/__tests__/test-delete.R [hunk 2]
```diff
@@ -0,0 +1,32 @@
+test_that("mc_delete_object works correctly with local TRUE", {
+  with_envvar(
+    new = c(HS_LOCAL = TRUE, MAILCHIMP_API_KEY = "ABC"), {
+      without_internet(
```
[the remark this answers is not in the document]
Comment (zackarno) [reply src/email/mailchimp/__tests__/test-delete.R:4]: but it does look like it is missing the `{}` syntax in the `without_internet()` documentation and as in next test below... so im not sure if it's actually working

**Comment by zackarno** [comment]
some checklists... because i want to check boxes

- [ ] run GHA dispatch on all `monitor_*` scripts with `HS_DRY_RUN =TRUE` and `HS_LOCAL= TRUE`
- [ ] run GHA dispatch on all `monitor_*` scripts with `HS_DRY_RUN =TRUE` and `HS_LOCAL= TRUE`

note this won't work on ACLED script until we merge the patch for the territory in Cyprus

File: src/email/mailchimp/__tests__/test-delete.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (martinig94) [reply src/email/mailchimp/__tests__/test-delete.R:4]: I can't replicate this behavior, let's check it together in our call later

**Review by martinig94 — APPROVED** [review]

**Pull request merged**